kythe/go/serving/xrefs/columnar/columnar_encoding.go

```diff
@@ -0,0 +1,366 @@
+/*
+ * Copyright 2018 The Kythe Authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Package columnar implements the columnar table format for a Kythe xrefs service.
+package columnar
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+
+	"kythe.io/kythe/go/util/keys"
+	"kythe.io/kythe/go/util/kytheuri"
+
+	"github.com/golang/protobuf/proto"
+
+	scpb "kythe.io/kythe/proto/schema_go_proto"
+	spb "kythe.io/kythe/proto/storage_go_proto"
+	xspb "kythe.io/kythe/proto/xref_serving_go_proto"
+)
+
+// ColumnarDecorationsKeyPrefix is the common key prefix for all Kythe columnar
+// FileDecoration key-value entries.
+var ColumnarDecorationsKeyPrefix, _ = keys.Append(nil, "fd")
+
+func init() {
+	// Restrict the capacity of the key prefix to ensure appending to it creates a new array.
+	ColumnarDecorationsKeyPrefix = ColumnarDecorationsKeyPrefix[:len(ColumnarDecorationsKeyPrefix):len(ColumnarDecorationsKeyPrefix)]
```

Comment:
what does slice[:N:N] do?

Reply:
It ensures that the capacity of the `[]byte` is the same as its length.  This means that any `append` operation will copy the underlying bytes instead of using any leftover capacity.  If we didn't do this, any concurrent use of the prefix would be a race.